### DISCUSSION TITLE
Documentation Structure Block Editor Handbook

### DISCUSSION THREAD

**Discussion by ghost** [WordPress/gutenberg#48998] [Developer Experience]
After discussing some ideas for the Block Editor Handbook with @ryanwelcher, we thought the best next step would be to open a discussion for everyone to join in.

First of all, thanks everyone who is working on the documentation. I know it's a huge effort to get everything right and that the ratio of the team size and the amount of work that needs to be done is not common. Every feedback I have is based on potential I'm seeing, not to criticize anything that's been done so far.

**The main documentation issue**
The main issue I have with the documentation is how everything is so scattered, which is a result of having such a modular system with many different NPM packages. The generated docs are fine as a reference, but they're all isolated and I think we need a place that explains how these different packages work together.

It also has to be in a more chronological way, instead of taking a single isolated subject and putting everything there is to know, from beginner info to the most advanced inner workings, all on one page.

**The Block Editor Handbook**
I think the Block Editor Handbook would be the right place for this so that will be the focus of this discussion, but to be just that I think there needs to be some change in it's structure.

**The Block Editor Handbook issues**
One issue is with consistency, I think we need a place where we can see what words we're using, if we use British English or American English and how we capitalize stuff. A few simple examples: I see `block` `Block` `block editor` `Block Editor` etc. all over the place. It's also confusing to call something the `Block Editor Handbook` and then start the first sentence with `Gutenberg`. Question a lot of new devs will have: What is Gutenberg and what is the Block Editor, are they the same or not? Why are we getting all this phases info about Gutenberg when learning about the Block Editor?

**The structure**
Right now the structure has these chapters:

<img width="240" alt="Screenshot 2023-03-10 at 5 43 54 PM" src="https://user-images.githubusercontent.com/118618474/224373358-66df4271-50da-4960-aec2-17dd11bcc20b.png">

Especially for non-native English speakers, this can be confusing. E.g. what's the difference between an explanation and a how-to? And why is getting started not a how to? I know there are nuances in these words, but they're not super clear for everyone.

I rewrote an outline for a new structure. The main goals for this restructure are to take away some confusion, pointing people to other places when there already is a documentation (DRY), and create a chronological handbook people can follow, but still give them the opportunity to dive into deeper subjects right away. 

```
- Introduction
- Prerequisites
    - Using React
    - Using Redux
    - Using NPM
    - Using PHP
- Getting Started
    - Development environment
    - Create a block (package)
- The Editor
    - Using the Editor
    - Extending the Editor
    - Site Editor vs. Block Editor
- Blocks
    - The anatomy of a block
        - Block Toolbar
        - Structure
        - Content
        - Sidebar
        - States
        - Setup
    - block.json
    - Block Attributes
    - Inner Blocks
    - Block Variations
    - Block Locking
    - Block Accessibility
    - Block Transforms
    - Block Patterns
    - Block Extensions
    - Dynamic vs. Static
- Reference
    - NPM Package Reference
    - Component Reference
    - Hooks Reference
        - Block Hooks
        - Editor Hooks
    - Core Blocks Reference
- Contributors (link to the guide on GitHub)
```

I would love to hear everyone's thoughts on this!

**Comment by StuurluiDevelopment** [comment]
@ryanwelcher @mburridge

**Comment by mburridge** [comment]
Thanks for this feedback @tomdevisser. You're absolutely right that the handbook should be more progressive in nature and that it should provide a gentler learning curve. Perhaps the [Introduction to Block Development course](https://learn.wordpress.org/course/introduction-to-block-development-build-your-first-custom-block/) over on learn.wordpress.org could be adapted to fit into the handbook in some way. Either a modified version of the course could replace the existing [Getting Started tutorial](https://developer.wordpress.org/block-editor/getting-started/create-block/), or the existing tutorial could be extended to cover more of the ground that the course covers.

Regarding the current structure of the handbook, it's loosely based on the [Unified Theory of Documentation](https://documentation.divio.com/). The structure you suggest is a great start. I particularly like the idea of having a section on "The Anatomy of a Block". That's something that I think is lacking in the current handbook. Although I'm sure the information is in there somewhere it's quite dissipated and hard to assimilate, so that's a great idea. A section on the anatomy of a block would, I guess, fall under "Explanations" in the Unified Theory.

I think the Unified Theory of Documentation leads to well structured docs that is multi-purpose and serves a variety of users and learners, so we should be cautious of deviating too much from that model. However, within the structure of that model I agree that there's a great deal of restructuring that can be done to improve the docs both from a learning point of view and from a reference point of view.

**Reply by ghost** [reply]
What is a good next step here? Mentioning it in doc meetings?

**Comment by johnhooks** [comment]
> The structure you suggest is a great start. I particularly like the idea of having a section on "The Anatomy of a Block". That's something that I think is lacking in the current handbook. 

As a developer new to Gutenberg / Block Editor (yeah, I don't know what to call it), the concept of "What is a Block" has taken a lot of exploration of the code base to figure out. I would also note the Block Attribute reference section could be a lot more detailed. I am very willing to help with this effort.

**Comment by mburridge** [comment]
Hey @tomdevisser, @johnhooks 

Thanks for your suggestions and offer of help.

I'm currently undertaking an analysis of the handbook with a view to exploring how best to re-organise and/or restructure the content. I'm going to do this in the form of a mindmap so that we have a visual representation of how the handbook is now, so it will be easier to conceptualise and visualise how it can be made better.

If you're interested in getting involved with this I'd be delighted to have you aboard. Perhaps we could each target different sections of the handbook - I'm currently working through the Getting Started section.

Another job that I think is important is to identify what are the key concepts and knowledge that a developer new to blocks needs to learn. It would be great to get a list of these key things together with a view to ordering them into a structured learning path or curriculum of learning.

What do you think?

**Reply by ghost** [reply]
Hi @mburridge! Thank you so much for working on this, really appreciate it. I think however, all starting with a section won't solve the current problem. Like I mentioned, I think the problem with the documentation right now is that there is a disconnection between sections, packages, and multiple places that have documentation. I feel like we need an overview first of what goes where, what sections we need and which we don't, and create a general outline before starting to work on sections on a lower level. Only then we can decide what information goes in what section, without having information duplicated or scattered in multiple places.

**Reply by mburridge** [reply]
I agree. I'm not suggesting working on sections at a lower level at this stage. The first thing to do, as I see it, is work on an overview of the current state of the documentation which will give us the perspective to determine what sections are needed and what aren't.

Following that the handbook can be "re-architected" - i.e. create a general outline, as you put it, of what the new structure would be.

And only **then** work on the content at a lower level.

So what I'm suggesting is not individually working on different sections, but individually working on **analysing** different sections so we can get that overview.

If you have any ideas for another strategy I'd be keen to hear them.

**Reply by johnhooks** [reply]
@tomdevisser I agree with developing an review of the handbook structure, though there are improvements that can be made in the current documents that would be a step in the right direction, especially the  Block API Reference. @mburridge I will start on a proposed revision of [Attributes](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-attributes/). I am currently deep into learning the inner workings of attributes in `@wordpress/blocks` and it is fresh in my mind.

**Reply by ghost** [reply]
@mburridge Ah, maybe I misunderstood, I thought you were already working on the Getting Started section separately from the others.

I would suggest to start by perfecting the structure tree I made in the initial post. See what has to be added/removed, which sections could be combined or split up, and finish that outline first.

Once we did that, we can analyze the current docs and see what text goes where in those sections. During this stage we can also see how much information is now duplicated in different sections and merge those together.

After that we can look at each section separately and see if some sections are missing info, or could have links to further reading. This would also be a good time to add examples where they're missing.

**Reply by mburridge** [reply]
> I thought you were already working on the Getting Started section separately from the others.

No, I'm working on an analysis, or mindmap, of the Getting Started section - though I intend to do this for all the sections. With this overview in place we'll be able to see where there is duplication, where there are omissions, etc...

I think starting with the structure tree you made initially would be a great approach if we were starting from scratch, and oh how I wish we could! Unfortunately that's not the case and we have a legacy of content to deal with and organise.

The other issue to think about when re-organising the content is 301 redirects - but that's a problem for later on in the process.

**Reply by ghost** [reply]
> if we were starting from scratch, and oh how I wish we could! Unfortunately that's not the case

Why not? :) I think that's exactly what's best here, starting from scratch structure wise. Of course, we wouldn't have to throw any content away, but we can reorganize is using a new and improved structure. I often wished they did this with WordPress earlier in the project, like most big CMSs, instead of only adding new stuff. With Gutenberg we're still relatively early, and I think this would be the perfect time to do it.

**Reply by ghost** [reply]
> No, I'm working on an analysis, or mindmap, of the Getting Started section - though I intend to do this for all the sections

And for this, wouldn't it be nice to start with a mindmap of all the sections we want, instead of making a mindmap for a specific section (like Getting Started), wouldn't that be a better place to start?

**Reply by mburridge** [reply]
>  I think that's exactly what's best here, starting from scratch structure wise.

That would be ideal but I think it's definitely something that neither you nor I can decide by ourselves. I would recommend that this suggestion be proposed in the [#core-editor](https://wordpress.slack.com/archives/C02QB2JS7) channel in Making WordPress Slack so we can get community engagement and involvement. It would certainly be interesting to see how much enthusiasm there was for starting over with the handbook, and how much pushback there was.

Do you wanna hop over there and stick your head above the parapet on this? 😅

Might be an idea to also get some feedback from [#docs](https://wordpress.slack.com/archives/C02RP4WU5) channel regarding this too.

**Reply by mburridge** [reply]
> wouldn't it be nice to start with a mindmap of all the sections we want

How do we determine all the sections we want. I think we need to review what we have first. I, for one, haven't read the entire handbook cover to cover (so to speak). I don't know everything that's in there, and I certainly don't know everything there is to know about block development, but I don't know what I don't know. 🤷‍♂️

**Reply by ghost** [reply]
> How do we determine all the sections we want

My question would be, how do we write a good and structured documentation if we don't have an answer to this? Of course we can add/move/merge sections as we go, but from experience documenting at my company I know we need a high-level overview before we start writing, that's why I made the outline. Orienting in the repository, online and in the Slack channel can help with filling in our knowledge gaps. Asking if we missed or misplaced something.

**Reply by ghost** [reply]
> Do you wanna hop over there and stick your head above the parapet on this? 😅

I definitely would as I really believe in this idea, but I would like to have some support on this so people take it seriously and won't discard it as a new contributor wanting too much. I know it's a big undertaking, but I'm willing to put in the effort in guiding it if I would be trusted with the responsibility. @ryanwelcher is this something you'd want to chime in on?

**Reply by mburridge** [reply]
> My question would be, how do we write a good and structured documentation if we don't have an answer to this?

We totally need an answer to this. I'm just suggesting preliminary steps that will help us arrive at a determination of the sections we need and the order that they should be in. Those preliminary steps being: seeing what we've already got and getting an overview. I think this is worth doing whether we revise what's there, or even if we go for the starting from scratch again option.

**Reply by ghost** [reply]
@mburridge Of course, by all means do what you think helps, especially if you've never read it back to back and you feel like you need a refresher on what's out there. I read it back to back 1.5 weeks ago, and that was definitely helpful. I only admire putting in the work right away instead of only talking about it. Just wanted to discuss what would be a good first step for the collaborating effort, instead of individually.

**Reply by mburridge** [reply]
> I would like to have some support on this so people take it seriously and won't discard it as a new contributor wanting too much

New contributors bring new ideas and innovative insights, and that's totally welcomed in the WP community. I'm always happy to help support a point of view that's clearly thought through and coherently argued for.

**Reply by mburridge** [reply]
> I read it back to back 1.5 weeks ago, and that was definitely helpful.

Wow, respect! 🙇‍♂️ I see no harm in pursuing both avenues. I'm happy to continue working on analysing the handbook, it also forces me to read pretty much all of it as I go. Meanwhile you could follow your avenue and propose a structure for a "starting from scratch" version - but I wouldn't spend too much time or effort on this without some community backing.

**Reply by ghost** [reply]
Sounds like a plan. If you feel like it, it would be nice if while reading you'd sometimes take a look at the structure outline I made and see if it makes sense to you or if you would make changes here and there. Doesn't have to take much time, just a glance every now and then would be great!

**Reply by mburridge** [reply]
> it would be nice if while reading you'd sometimes take a look at the structure outline I made

Absolutely I will. Is that your proposal, or is it a starting point that you intend to refine and develop further? It would be good to liaise together and work with each other. I'd totally be interested in seeing how your proposal develops and would be open to helping out wherever I can and offering feedback and suggestions in the light of the analysis of the existing content that I'm conducting.

One thing that I think is super important is that any new version of the handbook needs to be as complete and as comprehensive as possible. While it would be good to streamline the existing handbook by removing duplicated content, we should make sure that everything in the existing handbook is also in any new version. Ideally we should try to avoid what happened when the old WP Codex was updated to the newer documentation in that the newer version was done in a piecemeal fashion and a lot of important stuff was omitted at first, and so developers continued to refer to the Codex for quite a considerable time. In fact some developers still prefer the Codex over the newer documentation. 😅 That's why I consider it important to get an overview of what we have already first and hence why I'm doing an analysis of the existing content before doing anything else like working on a new structure.

Having a new version is appealing. It has that "fresh start" allure, and could avoid the problem with 301 redirects that a revamp of the existing handbook would necessitate. And like the old Codex, the old version could still be kept around for those that still wanted to use it.

TBH I'm open-minded as to whether we revamp the current version or whether we re-write it from the ground up. For the latter we would certainly need community support so go ahead and post in Making WordPress Slack. I'll certainly chime in there and lend my support. Might be worth posting a link to this discussion so anyone interested can catch up with what's been discussed so far.

**Reply by mburridge** [reply]
Also, as I said earlier we should be cautious of producing documentation that isn't modelled on the [Unified Theory of Documentation](https://documentation.divio.com/). The current handbook already follows that model of:

- Tutorials
- How-to Guides
- Explanation
- Reference

Though some stuff needs moving around, for example there are tutorials in the How-to Guides section, and I'm sure there's other stuff in the wrong place. Stuff can be changed, re-worked, and moved around, but largely within the strictures of the Unified Theory as they lead to good and well structured documentation.

**Reply by ghost** [reply]
> Is that your proposal, or is it a starting point that you intend to refine and develop further? It would be good to liaise together and work with each other.

A starting point, I'm sure we can improve it.

> we should make sure that everything in the existing handbook is also in any new version

Definitely. We shouldn't "publish" a 2.0 (although I would rather call it 1.0 and the current one a beta) version before it's an improvement.

> For the latter we would certainly need community support so go ahead and post in Making WordPress Slack.

Will do this on the next open floor!

> Also, as I said earlier we should be cautious of producing documentation that isn't modelled on the [Unified Theory of Documentation](https://documentation.divio.com/). The current handbook already follows that model of: tutorials, how-to guides (...) .

Although I think it's nice to have a reference, I don't think this should be regarded as our holy grail. Like with coding standards, we use common sense to make things more readable, and I immediately see a few fundamental flaws in the model of the "Unified Theory".

E.g. what is the difference between a tutorial and a how-to guide? And aren't they both explanations? And if I want to know how to do something, am I gonna scan the tutorials, explanations, how-to's or reference? Or all of them? It's not clear what goes where, which is why there's a lot of duplication and no way to see at a glance if something's missing.

**Reply by johnhooks** [reply]
Just from experience the tailwind.css, documentation is amazing. They actually make a point to repeat content so users don't have to back out of the current section to get context. @tomdevisser, I just want to throw it out there that duplication can actually improve the DX, as long as, it is relevant and correct. 

I'll also commit to reading the current version cover to cover and document pain points.

**Reply by juanmaguitar** [reply]
> I just want to throw it out there that duplication can actually improve the DX, as long as, it is relevant and correct.

While I agree it may improve DX, content duplication also brings maintenance issues that in the long term could introduce bigger DX problems (information not accurate or out of date).

**Reply by johnhooks** [reply]
@juanmaguitar I agree if the duplication is hand written. But I believe tailwind.css does it in code, so sections can be repeated, but not be a maintenance nightmare.

**Comment by estelaris** [comment]
My comments here from a design/UX perspective:

> A few simple examples: I see block Block block editor Block Editor etc. all over the place. It's also confusing to call something the Block Editor Handbook and then start the first sentence with Gutenberg. Question a lot of new devs will have: What is Gutenberg and what is the Block Editor, are they the same or not? Why are we getting all this phases info about Gutenberg when learning about the Block Editor?

I see several problems here. 

1. First identify when referring to a "block" as an item and when is "Block" as part of the name. The docs team has a [style guide ](https://make.wordpress.org/docs/style-guide/) for this purpose.
2. Second, we are using a more international English, not too British, not too American. What that means is that we try to remove regionalisms as much as possible (like capitalizing every word in a title, "every two weeks" vs "fortnight"), use of simple language (e.g. async vs asynchronous). The goal is to help non-English speakers comprehend better what is written.
3. Third, there needs to be a content review to bring the documentation up-to-date (for instance, how long are we going to use the term "Gutenberg"?), clarify terms (difference between Block Editor and Gutenberg), etc.

I am not a fan of the Unified Theory of Documentation because it is restrictive, it doesn't allow for growth and software development needs a different structure when it is in continuous development. The structure proposed by @tomdevisser is a good starting point. 

There are definitely no blockages in creating a new structure to the documentation if it is going to improve the DX. I suggest looking into the list of articles (documentation) available now and define where in the process each goes, which is pretty much what Tomas did.

**Reply by mburridge** [reply]
Hi @estelaris, thanks for sharing your thoughts and thanks for the link to the style guide.

> I am not a fan of the Unified Theory of Documentation because it is restrictive, it doesn't allow for growth 

I don't necessarily agree that it's restrictive and doesn't allow for growth. What you might perceive as restrictive could also be seen as categorisation of different types of documentation serving different purposes. I think that it allows for growth, but that growth is "siloed" according to type of documentation.

> I suggest looking into the list of articles (documentation) available now and define where in the process each goes, which is pretty much what Tomas did.

Are you referring here to @tomdevisser's proposed structure, or is this someone else called Tomas who has done some other work on this? If the latter can you point us to that work? I am currently undertaking an analysis of the handbook and creating a mindmap of it, so if there's some prior work on this it could be useful for me.

**Reply by ghost** [reply]
Yeah that's me @mburridge , my Slack/wordpress.org name is Thomas de Visser so I understand the confusion.

**Reply by mburridge** [reply]
Okay, cool. So how are we going to move things forward here? Apart from @estelaris there hasn't been any engagement as a result of amplifying this in the [#docs meeting yesterday](https://wordpress.slack.com/archives/C02RP4WU5/p1679407205002259). What are our next steps? And what are we thinking, update the handbook or complete re-write from the ground up?

**Reply by ghost** [reply]
Today's the core editor meeting. I'll mention it there, and after that I'll make a list with to-do's and we can assign them amongst everyone willing to help. Sounds good? And I think everyone so far has been in favor of re-writing as it helps DX. 

Once we have the list and contributors, it might be nice to do some short calls every now and then to see where we are and what needs to be done, and add a summary of each call on this discussion.

**Reply by mburridge** [reply]
Awesome @tomdevisser, sounds like a plan. 👌 Looking forward to collaborating with you on this.

**Reply by ghost** [reply]
I feel like everyone has had a chance to give their perspectives on this now, so let's move on to the next step. Let's plan a call/meeting?

**Reply by mburridge** [reply]
Absolutely! If you're interested in attending a zoom meeting, or a series of zoom meetings, to start discussing a new structure for the handbook and working on a revised version please indicate the timezone you're in so we can schedule something at a time that suits the most people.

I'm in UTC +1.

**Reply by mburridge** [reply]
I've also [posted](https://wordpress.slack.com/archives/C02QB2JS7/p1679998785365479) in #core-editor.

**Reply by fabiankaegy** [reply]
I'd also love to attend the chats if time permits. I'm in UTC +2

**Reply by ndiego** [reply]
Same, I am UTC -5

**Reply by johnhooks** [reply]
I will attend if my schedule allows. Though I am UTC -8 and understand if a time is chosen ideal to the majority.

**Reply by estelaris** [reply]
Happy to attend to, if I can be of any help. I have posted about this in #core and again in #docs. In my experience, peeps don't get too involved in the beginning but are great at giving feedback and suggesting changes once something is presented. Don't be discouraged, I can push on the changes from the docs side and get the right people involved as soon as we have something to show.
I'm in UTC +2

**Reply by mburridge** [reply]
@tomdevisser I presume you're in too? What timezone are you in?

**Reply by ghost** [reply]
Definitely! I'm in Amsterdam, UTC+2

**Reply by mburridge** [reply]
Cool, thanks all 👌. I'll get something scheduled for next week at a time that accommodates as many of you as possible - it will probably need to be early-mid afternoon in Europe and morning in the Americas. What day would suit you best? How does Tue 04 Apr suit, at say 15:00UTC?

**Reply by mburridge** [reply]
Hey all, [look below](https://github.com/WordPress/gutenberg/discussions/48998#discussioncomment-5508002) 👇

**Comment by mburridge** [comment]
I've set up a Zoom meeting to discuss restructuring / rewriting the handbook as per the forgoing discussion.

I've scheduled it for 15:00 UTC tomorrow 04 April 2023. Come with your amazing ideas, thoughts on what can be improved, and where you think there are gaps that need to be filled. All welcome. 

I look forward to seeing you all and moving this project forward.

---

Zoom 3 is inviting you to a scheduled Zoom meeting.

Topic: Block Editor Handbook structure
Time: Apr 4, 2023 03:00 PM Greenwich Mean Time

Join Zoom Meeting
https://us02web.zoom.us/j/84021525365?pwd=NC9uVHQyMHEySUQwUU5teExKbFJaUT09

Meeting ID: 840 2152 5365
Passcode: 671588
One tap mobile
+13863475053,,84021525365#,,,,*671588# US
+15074734847,,84021525365#,,,,*671588# US

Dial by your location
        +1 386 347 5053 US
        +1 507 473 4847 US
        +1 564 217 2000 US
        +1 646 931 3860 US
        +1 669 444 9171 US
        +1 669 900 6833 US (San Jose)
        +1 689 278 1000 US
        +1 719 359 4580 US
        +1 929 205 6099 US (New York)
        +1 253 205 0468 US
        +1 253 215 8782 US (Tacoma)
        +1 301 715 8592 US (Washington DC)
        +1 305 224 1968 US
        +1 309 205 3325 US
        +1 312 626 6799 US (Chicago)
        +1 346 248 7799 US (Houston)
        +1 360 209 5623 US
Meeting ID: 840 2152 5365
Passcode: 671588
Find your local number: https://us02web.zoom.us/u/kdRqVDwavu

**Reply by mburridge** [reply]
**Please note**: the meeting will be recorded.

**Reply by mburridge** [reply]
Apologies, please use this link to join the meeting: https://us02web.zoom.us/j/89687217223?pwd=Z01kTmNnWXlneEhMRXI0Y0lYL0llZz09

**Comment by mburridge** [comment]
I've uploaded the video from the meeting to YouTube. I don't think I'll have the time to write up the notes this side of the Easter break, but if someone else wants to have a go, feel free and post the notes as a reply to this.

[![Alt text](https://img.youtube.com/vi/1BQMS1eJ8A4/0.jpg)](https://www.youtube.com/watch?v=1BQMS1eJ8A4)

**Reply by ghost** [reply]
I'm very sorry for not attending, don't know if Michael shared it, but my dad got diagnosed with cancer this week so I've been working less and have been in another headspace. I firstly wanted to thank you all for holding this meeting and making a great start towards improving in my opinion one of the most important parts of the project. I'm watching the video right now and will finish it today so I'm up to speed with what has been discussed. Might drop some comments here later!

**Reply by ghost** [reply]
Okay, watched the entire thing. First off, I think there's some really good ideas and insights here. It's super nice to see everyone has good ideas, and there seems to be a very much aligned vision on where we need to go, so that's great.

Having said that, I think the next meeting could use an approach that's a bit more pragmatic. There's been a lot of great ideas on content and pain points, but throwing them out there when we're not at a point where we can directly address them yet could lead to forgetting to address them at a later point, which would be a shame.

I think we need to divide this project into 3 stages.

1. Creating the structure (discussion overall hierarchy/structure)
2. Expanding the outline (discussion per subject)
3. Writing the content (discussion on low-level details)

For the next meeting I would suggest setting the goal of finishing the outline structure that I started in the initial post. Someone could share their screen and we can go over the structure and discuss what needs to be added/(re)moved/changed. This can still be changed, but it would be a 1.0 concept that we can then start working on. That would complete stage 1.

Once we have this actual outline, we can start making an outline per page/chapter. For example if we have the outline as mentioned above, we'd start with `Introduction`. What has to be on this page? What shouldn't be there? We can make a short list of questions we could ask ourselves and then go through all the chapters, `Prerequisites`, `Getting started`, etc. This is also the part where we can discuss all the problems and ideas that came up in this meeting, when we're actually at the chapter it relates to. This means once an idea or problem is mentioned we can come to a solution right away, instead of mentioning it and then thinking "yes we need to address that at some point". This would complete stage 2.

At this point we'll have a structure outline, plus a description of what should be on that page. This means we can start writing the actual content! We'll go over a few revisions and discussions once we have some content to discuss, and when this stage is done we have a perfectly thought out handbook. This would complete stage 3.

**Reply by ghost** [reply]
@ndiego @mburridge @estelaris @fabiankaegy @carolinan Please see my comments above :) (couldn't find Jenni's GitHub handle, if someone knows it please include her here as well)

**Reply by ghost** [reply]
One last addition to this, regarding the compartmentalizing suggestion Michael makes at the ending, I would highly recommend waiting with this until stage 2, and after every substep in stage 2 going back to stage 1 to see if the outline still works or if it could be improved. The biggest cause of most issues here, but also in WordPress (in my opinion), right now is compartmentalizing everything from the start and losing the bigger picture. Let's take a top down approach before working on low-level issues. I also wouldn't go async until the content writing stage.

**Reply by estelaris** [reply]
@jennimckinnon see above

**Comment by johnhooks** [comment]
My initial thoughts after watching the discussion:

- I agree the documentation needs to be versioned. Especially with the complexity of compatibility with Gutenberg and WordPress versions.
- There needs to be a distinction of the intended audience. I see at least ~three~ two:
  - ~General Audience - Non-technical users looking for usage information about the block editor.~
  - Plugin Developer - Implementing of block plugins, needing documentation for a specific version of WordPress.
  - Gutenberg Developer - Working directly on the source code of the editor, needing the latest version of Gutenberg.
- Who is this document for? Can a single document actually target both audiences?
-  If it is versioned, how can it be signaled to the reader they may be looking for documentation in the wrong location?
- As someone new to WordPress the fragmentation of documentation has been a struggle. Its compound by the way Gutenberg is published in packages versioned totally separate from the Gutenberg package. It's not at all apparent when beginning block development and installing `@wordpress/components` into a project, the version indicated in the `package.json` has no effect on the build asset. It enqueues `wp-components` which will be whatever version was included in release of WordPress that load it, or Gutenberg if installed.

As a developer, I am most interesting in a detailed, versioned, interlinked, and searchable API reference (anything like the [Code Reference](https://developer.wordpress.org/reference/) would be a huge improvement). To make sure it stays up to date I would think it needs to be part of the release cycle and most of it generated from JSDoc comments.

**Reply by estelaris** [reply]
@johnhooks the documentation in this handbook is focused on developers, in general and Gutenberg/Block developers in particular. Although we cannot hold anyone from reading the documentation, it is not written for a general-non-technical audience. There is the end-user documentation for the general-non-technical audience.

**Reply by johnhooks** [reply]
Thank you @estelaris I struck general audience.

**Comment by johnhooks** [comment]
I understand there is a difference between a good API reference and the Handbook, but the handbook is where the current Block API reference is located. Like @ndiego and @fabiankaegy, I find myself looking through the Gutenberg repository for basic information which should be in the reference. Thought because the README.md files are not interlinked, it can be a struggle to jump from the documentation of one package to the next. It also requires checking out the appropriate git release branch to make sure the documentation and code is from a specific version.

Should pursuing this goal be separate from the Handbook, just in the same way the WordPress Code Reference is independent from its Handbooks?

**Comment by estelaris** [comment]
I am adding the following comment from @alexstine in here, as a reference for the changes we are suggesting, he has good points regarding accessibility that will be handy when we review content

> I just wish blocks for developers were documented better. I do not find it useful that I need to adjust the block in the editor and then copy the code into a block theme. Why not instead tell me which blocks have which attributes and let me adjust it myself? There is a dependency on the block editor that makes development harder and the evolution of this concept is kind of scary. Why create block themes at all when in reality, the editor does all the work? It will lead to an impressive set of problems down the road as users getting near certain areas of websites never ends well. I like the idea of being able to output certain blocks by typing the HTML-like code because it is a lot of how PHP still works. Call the function, outputs some value. That is simple enough to understand. Docs in most places around core functions are good and if not, the source code is viewable right on codex.
Related doc: https://developer.wordpress.org/themes/block-themes/templates-and-template-parts/#how-to-find-which-block-markup-to-use
Anyone else find this lacking? Of course for me, this is slightly more annoying due to the accessibility of some of these blocks. I just figured for all devs this may get annoying after a while to require the block editor to do what could be done in minutes with PHP/HTML. Might be an unpopular opinion but I guess I am kind of good at that lately.
Compared with: https://developer.wordpress.org/reference/functions/get_the_title/

> It is also worth saying, developing Gutenberg on Windows is possible. I do it. You can use WSL for all the Linux parts. Handles Node.js a lot better than Windows Node.js installer. If it is just a case of having older Mac, I can see where that may be a problem. Building the project takes some resources. I prefer building on my Mac but all hope is not lost for our Windows users. I use both for dev.

**Comment by alexstine** [comment]
Adding to this, it is not an accessibility concern alone as much as it is a development concern. No developer wants to waste time in the block editor when they could simply open the docs and know exactly how to build the proper JSON structure in templates. Gutenberg is a JSON generator, it is what it does in practice. That does not mean there should not be docs around what each attribute does. WordPress is really how I learned PHP because the project is documented so well. I hate to see Gutenberg move us away from helpful learning materials.

**Reply by ndiego** [reply]
Hi @alexstine, I want to make sure I understand correctly. Using the Columns block as an example, the block includes the option to disable stacking on mobile. If toggled, the attribute `"isStackedOnMobile":false` is added to the block. Are you referring to better documentation on how each of these attributes work? 

The [Core Block Reference](https://developer.wordpress.org/block-editor/reference-guides/core-blocks/) does indicate all of the custom attributes for core blocks, but it doesn't go into any detail beyond just listing the attributes.

**Comment by alexstine** [comment]
@ndiego Yes, that is exactly what I am after. Something with details. Take this example page.

https://developer.wordpress.org/reference/functions/get_the_title/

Now, instead of PHP, think about how this format could work for each block.

Columns block
Description of the block.
Documented block attributes
Examples
Source Code on GitHub

Having everything on 1 page also feels a little overly messy to me especially given how the project will progress in the future.

Thanks.

**Comment by carolinan** [comment]
Especially for the block attributes, the current reference does not give enough information; it doesn't even say if it's a string or a boolean, etc. It doesn't give an example of how or even if that attribute is written in the block markup.
